The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11060670020/job/30731691051) [(plain)](https://github.com/rust-lang/rust/commit/912e1ae023587d9b1aabe71412c4239d22ffb200/checks/30731691051/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.773 Building wheels for collected packages: reuse
#13 2.774   Building wheel for reuse (pyproject.toml): started
#13 3.019   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.021   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 3.021   Stored in directory: /tmp/pip-ephem-wheel-cache-69jvah8r/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.024 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.414 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.415 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 3.949 Collecting virtualenv
#13 3.949 Collecting virtualenv
#13 4.001   Downloading virtualenv-20.26.5-py3-none-any.whl (6.0 MB)
#13 4.253      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.0/6.0 MB 24.1 MB/s eta 0:00:00
#13 4.298 Collecting distlib<1,>=0.3.7
#13 4.307   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 4.318      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 48.9 MB/s eta 0:00:00
#13 4.356 Collecting platformdirs<5,>=3.9.1
#13 4.364   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.403 Collecting filelock<4,>=3.12.2
#13 4.411   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.491 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.676 Successfully installed distlib-0.3.8 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.26.5
#13 DONE 4.8s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      192448 kB
DirectMap2M:     9244672 kB
DirectMap1G:     9437184 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://static.rust-lang.org/dist/2024-09-22/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/2024-09-22/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt
---
##[endgroup]
fmt check
fmt: checked 5594 files
tidy check
tidy error: `/checkout/src/librustdoc/html/templates/source.html` at line 12: unneeded `{# #}` tag at the end of the line
tidy error: `/checkout/src/librustdoc/html/templates/print_item.html` at line 1: unneeded `{# #}` tag at the end of the line
removing old virtual environment
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.2)
All checks passed!
checking C++ file formatting
some tidy checks failed
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Thu Sep 26 21:46:36 UTC 2024
  network time: Thu, 26 Sep 2024 21:46:36 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/src/librustdoc/html/templates/item_info.html
+++ b/src/librustdoc/html/templates/item_info.html
@@ -1,7 +1,7 @@
 {% if !items.is_empty() %}
     <span class="item-info">
         {% for item in items %}
-            {{item|safe}} {# #}
+            {{item|safe}}
         {% endfor %}
     </span>
 {% endif %}

--- a/src/librustdoc/html/templates/page.html
+++ b/src/librustdoc/html/templates/page.html
@@ -12,13 +12,13 @@
     <link rel="stylesheet" {#+ #}
           href="{{static_root_path|safe}}{{files.normalize_css}}"> {# #}
     <link rel="stylesheet" {#+ #}
-          href="{{static_root_path|safe}}{{files.rustdoc_css}}"> {# #}
+          href="{{static_root_path|safe}}{{files.rustdoc_css}}">
     {% if !layout.default_settings.is_empty() %}
     <script id="default-settings" {#+ #}
       {%~ for (k, v) in layout.default_settings ~%}
         data-{{k}}="{{v}}"
       {% endfor %}
-    ></script> {# #}
+    ></script>
     {% endif %}
     <meta name="rustdoc-vars" {#+ #}
          data-root-path="{{page.root_path|safe}}" {#+ #}
@@ -31,36 +31,36 @@
          data-search-js="{{files.search_js}}" {#+ #}
          data-settings-js="{{files.settings_js}}" {#+ #}
     > {# #}
-    <script src="{{static_root_path|safe}}{{files.storage_js}}"></script> {# #}
+    <script src="{{static_root_path|safe}}{{files.storage_js}}"></script>
     {% if page.css_class.contains("crate") %}
-    <script defer src="{{page.root_path|safe}}crates{{page.resource_suffix}}.js"></script> {# #}
+    <script defer src="{{page.root_path|safe}}crates{{page.resource_suffix}}.js"></script>
     {% else if page.css_class == "src" %}
     <script defer src="{{static_root_path|safe}}{{files.src_script_js}}"></script> {# #}
-    <script defer src="{{page.root_path|safe}}src-files{{page.resource_suffix}}.js"></script> {# #}
+    <script defer src="{{page.root_path|safe}}src-files{{page.resource_suffix}}.js"></script>
     {% else if !page.css_class.contains("mod") %}
-    <script defer src="sidebar-items{{page.resource_suffix}}.js"></script> {# #}
+    <script defer src="sidebar-items{{page.resource_suffix}}.js"></script>
     {% else if !page.css_class.contains("sys") %}
-    <script defer src="../sidebar-items{{page.resource_suffix}}.js"></script> {# #}
+    <script defer src="../sidebar-items{{page.resource_suffix}}.js"></script>
     {% endif %}
-    <script defer src="{{static_root_path|safe}}{{files.main_js}}"></script> {# #}
+    <script defer src="{{static_root_path|safe}}{{files.main_js}}"></script>
     {% if layout.scrape_examples_extension %}
-    <script defer src="{{static_root_path|safe}}{{files.scrape_examples_js}}"></script> {# #}
+    <script defer src="{{static_root_path|safe}}{{files.scrape_examples_js}}"></script>
     {% endif %}
     <noscript> {# #}
         <link rel="stylesheet" {#+ #}
            href="{{static_root_path|safe}}{{files.noscript_css}}"> {# #}
-    </noscript> {# #}
+    </noscript>
     {% if layout.css_file_extension.is_some() %}
         <link rel="stylesheet" {#+ #}
-            href="{{page.root_path|safe}}theme{{page.resource_suffix}}.css"> {# #}
+            href="{{page.root_path|safe}}theme{{page.resource_suffix}}.css">
     {% endif %}
     {% if !layout.favicon.is_empty() %}
-        <link rel="icon" href="{{layout.favicon}}"> {# #}
+        <link rel="icon" href="{{layout.favicon}}">
     {% else %}
         <link rel="alternate icon" type="image/png" {#+ #}
             href="{{static_root_path|safe}}{{files.rust_favicon_png_32}}"> {# #}
         <link rel="icon" type="image/svg+xml" {#+ #}
-            href="{{static_root_path|safe}}{{files.rust_favicon_svg}}"> {# #}
+            href="{{static_root_path|safe}}{{files.rust_favicon_svg}}">
     {% endif %}
     {{ layout.external_html.in_header|safe }}
 </head> {# #}
@@ -69,60 +69,60 @@
     <div class="warning"> {# #}
         This old browser is unsupported and will most likely display funky things. {# #}
     </div> {# #}
-    <![endif]--> {# #}
+    <![endif]-->
     {{ layout.external_html.before_content|safe }}
     {% if page.css_class != "src" %}
     <nav class="mobile-topbar"> {# #}
-        <button class="sidebar-menu-toggle" title="show sidebar"></button> {# #}
+        <button class="sidebar-menu-toggle" title="show sidebar"></button>
         {% if !layout.logo.is_empty() || page.rust_logo %}
-        <a class="logo-container" href="{{page.root_path|safe}}{{display_krate_with_trailing_slash|safe}}index.html"> {# #}
+        <a class="logo-container" href="{{page.root_path|safe}}{{display_krate_with_trailing_slash|safe}}index.html">
         {% if page.rust_logo %}
-            <img class="rust-logo" src="{{static_root_path|safe}}{{files.rust_logo_svg}}" alt=""> {# #}
+            <img class="rust-logo" src="{{static_root_path|safe}}{{files.rust_logo_svg}}" alt="">
         {% else if !layout.logo.is_empty() %}
-            <img src="{{layout.logo}}" alt=""> {# #}
+            <img src="{{layout.logo}}" alt="">
         {% endif %}
-        </a> {# #}
+        </a>
         {% endif %}
     </nav>
     {% endif %}
-    <nav class="sidebar"> {# #}
+    <nav class="sidebar">
         {% if page.css_class != "src" %}
         <div class="sidebar-crate">
             {% if !layout.logo.is_empty() || page.rust_logo %}
-            <a class="logo-container" href="{{page.root_path|safe}}{{display_krate_with_trailing_slash|safe}}index.html"> {# #}
+            <a class="logo-container" href="{{page.root_path|safe}}{{display_krate_with_trailing_slash|safe}}index.html">
                 {% if page.rust_logo %}
-                    <img class="rust-logo" src="{{static_root_path|safe}}{{files.rust_logo_svg}}" alt="logo"> {# #}
+                    <img class="rust-logo" src="{{static_root_path|safe}}{{files.rust_logo_svg}}" alt="logo">
                 {% else if !layout.logo.is_empty() %}
-                    <img src="{{layout.logo}}" alt="logo"> {# #}
+                    <img src="{{layout.logo}}" alt="logo">
                 {% endif %}
-            </a> {# #}
+            </a>
             {% endif %}
             <h2> {# #}
-                <a href="{{page.root_path|safe}}{{display_krate_with_trailing_slash|safe}}index.html">{{display_krate|wrapped|safe}}</a> {# #}
+                <a href="{{page.root_path|safe}}{{display_krate_with_trailing_slash|safe}}index.html">{{display_krate|wrapped|safe}}</a>
                 {% if !display_krate_version_number.is_empty() %}
                     <span class="version">{{+ display_krate_version_number}}</span>
                 {% endif %}
             </h2> {# #}
-        </div> {# #}
+        </div>
         {% if !display_krate_version_extra.is_empty() %}
-        <div class="version">{{+ display_krate_version_extra}}</div> {# #}
+        <div class="version">{{+ display_krate_version_extra}}</div>
         {% endif %}
         {% else %}
         <div class="src-sidebar-title"> {# #}
             <h2>Files</h2> {# #}
-        </div> {# #}
+        </div>
         {% endif %}
         {{ sidebar|safe }}
     </nav> {# #}
     <div class="sidebar-resizer"></div> {# #}
-    <main> {# #}
+    <main>
         {% if page.css_class != "src" %}<div class="width-limiter">{% endif %}
             {# defined in storage.js to avoid duplicating complex UI across every page #}
             {# and because the search form only works if JS is enabled anyway #}
             <rustdoc-search></rustdoc-search> {# #}
-            <section id="main-content" class="content">{{ content|safe }}</section> {# #}
+            <section id="main-content" class="content">{{ content|safe }}</section>
         {% if page.css_class != "src" %}</div>{% endif %}
-    </main> {# #}
+    </main>
     {{ layout.external_html.after_content|safe }}
 </body> {# #}
 </html> {# #}

--- a/src/librustdoc/html/templates/print_item.html
+++ b/src/librustdoc/html/templates/print_item.html
@@ -1,4 +1,4 @@
-<div class="main-heading"> {# #}
+<div class="main-heading">
     {% if !path_components.is_empty() %}
     <span class="rustdoc-breadcrumbs">
         {% for (i, component) in path_components.iter().enumerate() %}

--- a/src/librustdoc/html/templates/short_item_info.html
+++ b/src/librustdoc/html/templates/short_item_info.html
@@ -3,21 +3,21 @@
         <div class="stab deprecated"> {# #}
             <span class="emoji">👎</span> {# #}
             <span>{{message|safe}}</span> {# #}
-        </div> {# #}
+        </div>
     {% when Self::Unstable with { feature, tracking } %}
         <div class="stab unstable"> {# #}
             <span class="emoji">🔬</span> {# #}
             <span> {# #}
                 This is a nightly-only experimental API. ({# #}
-                <code>{{feature}}</code> {# #}
+                <code>{{feature}}</code>
                 {% match tracking %}
                     {% when Some with ((url, num)) %}
-                        &nbsp;<a href="{{url}}{{num}}">#{{num}}</a> {# #}
+                        &nbsp;<a href="{{url}}{{num}}">#{{num}}</a>
                     {% when None %}
                 {% endmatch %}
                 ) {# #}
             </span> {# #}
-        </div> {# #}
+        </div>
     {% when Self::Portability with { message } %}
-        <div class="stab portability">{{message|safe}}</div> {# #}
+        <div class="stab portability">{{message|safe}}</div>
 {% endmatch %}

--- a/src/librustdoc/html/templates/sidebar.html
+++ b/src/librustdoc/html/templates/sidebar.html
@@ -30,7 +30,7 @@
                                             {% else %}
                                                 {{link.name}}
                                         {% endmatch %}
-                                    </a> {# #}
+                                    </a>
                                     {% if !link.children.is_empty() %}
                                         <ul>
                                             {% for child in link.children %}

--- a/src/librustdoc/html/templates/source.html
+++ b/src/librustdoc/html/templates/source.html
@@ -9,7 +9,7 @@
 </div>
 {% else %}
 {% endmatch %}
-<div class="example-wrap"> {# #}
+<div class="example-wrap">
     {# https://developers.google.com/search/docs/crawling-indexing/robots-meta-tag#data-nosnippet-attr
        Do not show "1 2 3 4 5 ..." in web search results. #}
     <div data-nosnippet><pre class="src-line-numbers">

--- a/src/librustdoc/html/templates/type_layout.html
+++ b/src/librustdoc/html/templates/type_layout.html
@@ -1,7 +1,7 @@
 <h2 id="layout" class="section-header"> {# #}
     Layout<a href="#layout" class="anchor">§</a> {# #}
 </h2> {# #}
-<div class="docblock"> {# #}
+<div class="docblock">
     {% match type_layout_size %}
     {% when Ok(type_layout_size) %}
     <div class="warning"> {# #}
@@ -14,19 +14,19 @@
             chapter for details on type layout guarantees. {# #}
         </p> {# #}
     </div> {# #}
-    <p><strong>Size:</strong> {{+ type_layout_size|safe }}</p> {# #}
+    <p><strong>Size:</strong> {{+ type_layout_size|safe }}</p>
     {% if !variants.is_empty() %}
     <p> {# #}
         <strong>Size for each variant:</strong> {# #}
     </p> {# #}
-    <ul> {# #}
+    <ul>
         {% for (name, layout_size) in variants %}
         <li> {# #}
             <code>{{ name }}</code>: {#+ #}
             {{ layout_size|safe }}
-        </li> {# #}
+        </li>
         {% endfor %}
-    </ul> {# #}
+    </ul>
     {% endif %}
     {# This kind of layout error can occur with valid code, e.g. if you try to
     get the layout of a generic type such as `Vec<T>`. #}
@@ -35,7 +35,7 @@
             <strong>Note:</strong> Unable to compute type layout, {#+ #}
             possibly due to this type having generic parameters. {#+ #}
             Layout can only be computed for concrete, fully-instantiated types. {# #}
-        </p> {# #}
+        </p>
         {# This kind of error probably can't happen with valid code, but we don't
         want to panic and prevent the docs from building, so we just let the
         user know that we couldn't compute the layout. #}
@@ -43,21 +43,21 @@
         <p> {# #}
             <strong>Note:</strong> Encountered an error during type layout; {#+ #}
             the type was too big. {# #}
-        </p> {# #}
+        </p>
         {% when Err(LayoutError::ReferencesError(_)) %}
         <p> {# #}
             <strong>Note:</strong> Encountered an error during type layout; {#+ #}
             the type references errors. {# #}
-        </p> {# #}
+        </p>
         {% when Err(LayoutError::NormalizationFailure(_, _)) %}
         <p> {# #}
             <strong>Note:</strong> Encountered an error during type layout; {#+ #}
             the type failed to be normalized. {# #}
-        </p> {# #}
+        </p>
         {% when Err(LayoutError::Cycle(_)) %}
         <p> {# #}
             <strong>Note:</strong> Encountered an error during type layout; {#+ #}
             the type's layout depended on the type's layout itself. {# #}
-        </p> {# #}
+        </p>
         {% endmatch %}
 </div> {# #}

--- a/src/tools/tidy/src/rustdoc_templates.rs
+++ b/src/tools/tidy/src/rustdoc_templates.rs
@@ -20,7 +20,39 @@ pub fn check(librustdoc_path: &Path, bad: &mut bool) {
 
             while let Some((pos, line)) = lines.next() {
                 let line = line.trim();
-                if TAGS.iter().any(|(_, tag)| line.ends_with(tag)) {
+                if let Some(need_next_line_check) = TAGS.iter().find_map(|(tag, end_tag)| {
+                    // We first check if the line ends with a jinja tag.
+                    if !line.ends_with(end_tag) {
+                        return None;
+                    // Then we check if this a comment tag.
+                    } else if *tag != "{#" {
+                        return Some(false);
+                    // And finally we check if the comment is empty (ie, only there to strip
+                    // extra whitespace characters).
+                    } else if let Some(start_pos) = line.rfind(tag) {
+                        Some(line[start_pos + 2..].trim() == "#}")
+                    } else {
+                        Some(false)
+                    }
+                }) {
+                    // All good, the line is ending is a jinja tag. But maybe this tag is useless
+                    // if the next line starts with a jinja tag as well!
+                    //
+                    // However, only (empty) comment jinja tags are concerned about it.
+                    if need_next_line_check
+                        && lines.peek().is_some_and(|(_, next_line)| {
+                            let next_line = next_line.trim_start();
+                            TAGS.iter().any(|(tag, _)| next_line.starts_with(tag))
+                        })
+                    {
+                        // It seems like ending this line with a jinja tag is not needed after all.
+                        tidy_error!(
+                            bad,
+                            "`{}` at line {}: unneeded `{{# #}}` tag at the end of the line",
+                            path.path().display(),
+                            pos + 1,
+                        );
+                    }
                     continue;
                 }
                 let Some(next_line) = lines.peek().map(|(_, next_line)| next_line.trim()) else {

